node/src/components/event_stream_server/tests.rs

```diff
@@ -0,0 +1,899 @@
+                client_behavior.barrier.wait().await;
+                debug!("server finished waiting before event {}", id);
+            }
+        }
+    }
+
+    /// Sleeps if `self` was set to enable delays between events.
+    async fn sleep_if_required(&self) {
+        if self.has_delay_between_events {
+            time::sleep(DELAY_BETWEEN_EVENTS).await;
+        }
+    }
+}
+
+/// A helper to allow the server to be kept alive until a specific call to stop it.
+#[derive(Clone)]
+struct ServerStopper {
+    should_stop: Arc<AtomicBool>,
+    notifier: Arc<Notify>,
+}
+
+impl ServerStopper {
+    fn new() -> Self {
+        ServerStopper {
+            should_stop: Arc::new(AtomicBool::new(false)),
+            notifier: Arc::new(Notify::new()),
+        }
+    }
+
+    /// Returns whether the server should stop now or not.
+    fn should_stop(&self) -> bool {
+        self.should_stop.load(Ordering::SeqCst)
+    }
+
+    /// Waits until the server should stop.
+    async fn wait(&self) {
+        while !self.should_stop() {
+            self.notifier.notified().await;
+        }
+    }
+
+    /// Tells the server to stop.
+    fn stop(&self) {
+        self.should_stop.store(true, Ordering::SeqCst);
+        self.notifier.notify_one();
+    }
+}
+
+impl Drop for ServerStopper {
+    fn drop(&mut self) {
+        self.stop();
+    }
+}
+
+struct TestFixture {
+    storage_dir: TempDir,
+    protocol_version: ProtocolVersion,
+    events: Vec<SseData>,
+    first_event_id: Id,
+    server_join_handle: Option<JoinHandle<()>>,
+    server_stopper: ServerStopper,
+}
+
+impl TestFixture {
+    /// Constructs a new `TestFixture` including `EVENT_COUNT` random events ready to be served.
+    fn new(rng: &mut TestRng) -> Self {
+        let _ = logging::init();
+        let storage_dir = tempfile::tempdir().unwrap();
+        fs::create_dir_all(&storage_dir).unwrap();
+        let protocol_version = ProtocolVersion::from_parts(1, 2, 3);
+
+        let events = (0..EVENT_COUNT)
+            .map(|_| match rng.gen_range(0..8) {
+                0 => SseData::random_block_added(rng),
+                1 => SseData::random_deploy_processed(rng),
+                2 => SseData::random_fault(rng),
+                3 => SseData::random_step(rng),
+                4..=7 => SseData::random_finality_signature(rng),
+                _ => unreachable!(),
+            })
+            .collect();
+
+        TestFixture {
+            storage_dir,
+            protocol_version,
+            events,
+            first_event_id: 0,
+            server_join_handle: None,
+            server_stopper: ServerStopper::new(),
+        }
+    }
+
+    /// Creates a new `EventStreamServer` and runs it in a tokio task, returning the actual address
+    /// the server is listening on.
+    ///
+    /// Only one server can be run at a time; this panics if there is already a server task running.
```

Comment:
Are we limiting ourselves to one instance (and thus one test in parallel) due to port collision issues?

Reply:
No, purely for simplicity in the tests.  The separate unit tests all run in parallel, but I couldn't see a use for running two servers in parallel in a single test.

Reply:
Well, it's just tests, so I am fine with that :)